src/server/electron.ts

```diff
@@ -0,0 +1,207 @@
+      // Resolving the race between the debugger and the boot-time script.
+      if ((global as any)._playwrightRun)
+        return (global as any)._playwrightRun();
+      return new Promise(f => (global as any)._playwrightRunCallback = f);
+    });
+  }
+
+  async evaluate<R, Arg>(pageFunction: types.FuncOn<any, Arg, R>, arg: Arg): Promise<R>;
+  async evaluate<R>(pageFunction: types.FuncOn<any, void, R>, arg?: any): Promise<R>;
+  async evaluate<R, Arg>(pageFunction: types.FuncOn<any, Arg, R>, arg: Arg): Promise<R> {
+    return this._nodeElectronHandle!.evaluate(pageFunction, arg);
+  }
+
+  async evaluateHandle<R, Arg>(pageFunction: types.FuncOn<any, Arg, R>, arg: Arg): Promise<types.SmartHandle<R>>;
+  async evaluateHandle<R>(pageFunction: types.FuncOn<any, void, R>, arg?: any): Promise<types.SmartHandle<R>>;
+  async evaluateHandle<R, Arg>(pageFunction: types.FuncOn<any, Arg, R>, arg: Arg): Promise<types.SmartHandle<R>> {
+    return this._nodeElectronHandle!.evaluateHandle(pageFunction, arg);
+  }
+
+  protected _computeDeadline(options?: types.TimeoutOptions): number {
+    return this._timeoutSettings.computeDeadline(options);
+  }
+}
+
+export class Electron  {
+  async launch(executablePath: string, options: ElectronLaunchOptions = {}): Promise<ElectronApplication> {
+    const {
+      args = [],
+      env = process.env,
+      handleSIGINT = true,
+      handleSIGTERM = true,
+      handleSIGHUP = true,
+      timeout = 30000,
+    } = options;
+    const deadline = helper.monotonicTime() + timeout;
+
+    const logger = new RootLogger(options.logger);
+    const electronArguments = ['--inspect=0', '--remote-debugging-port=0', '--require', path.join(__dirname, 'electronLoader.js'), ...args];
+    const { launchedProcess, gracefullyClose } = await launchProcess({
+      executablePath,
+      args: electronArguments,
+      env,
+      handleSIGINT,
+      handleSIGTERM,
+      handleSIGHUP,
+      logger,
+      pipe: true,
+      cwd: options.cwd,
+      omitDownloads: true,
+      attemptToGracefullyClose: () => app!.close(),
+      onkill: (exitCode, signal) => {
+        if (app)
+          app.emit(ElectronEvents.ElectronApplication.Close, exitCode, signal);
+      },
+    });
+
+    let app: ElectronApplication | undefined = undefined;
+    let browserServer: BrowserServer | undefined = undefined;
+
+    const timeoutError = new TimeoutError(`Timed out while trying to connect to Electron!`);
+    const nodeMatch = await waitForLine(launchedProcess, launchedProcess.stderr, /^Debugger listening on (ws:\/\/.*)$/, helper.timeUntilDeadline(deadline), timeoutError);
+    const nodeConnection = await WebSocketTransport.connect(nodeMatch[1], transport => {
+      return new CRConnection(transport, logger);
+    });
+
+    const chromeMatch = await waitForLine(launchedProcess, launchedProcess.stderr, /^DevTools listening on (ws:\/\/.*)$/, helper.timeUntilDeadline(deadline), timeoutError);
```

Comment:
Do we know that chrome match always comes later?

Reply:
Why does this matter?

Reply:
Yes!